### ISSUE TITLE
Adding properties to basic types

### ISSUE THREAD

**Issue by kikocorreoso** [brython-dev/brython#115]
Is there a canonical way to add a property to a python type implemented in js?

For instance, yesterday I was trying to add some (Python) properties to number classes (*py_float.js, py_int.js, py_complex.js*) using the [builtin `property`](https://github.com/brython-dev/brython/blob/master/src/py_builtin_functions.js#L1139) and the following code wasn't working:

```javascript
$IntDict.numerator = property(function(self){return self})
$IntDict.denominator = property(function(){return 1})
```

When I try to use the properties:
```python
a = 2
print(a.numerator)
```
 I get:
```
<property object>
```
Instead of the expected value `2`.

I also tried to use js builtin way to define properties:

```javascript
Object.defineProperty($IntDict, "numerator", {
    get: function(self) {return self},
    set: undefined
});
```

And the result was something like `int doesn't have attribute 'numerator'`.

**Comment by PierreQuentel** [comment]
Kiko,

This is because attribute resolution for built-in types is simplified, supposedly to improve performance. If you comment the attribute `$native=true` of `$IntDict` your code will produce the expected result.

I did a few quick performance tests to see if the use of `$native` really improves speed, it is not obvious. Could you also try with a few examples ? If it doesn't make much difference then we can remove it and you will be able to add these extra properties.

**Comment by kikocorreoso** [comment]
Thanks @PierreQuentel 

I will test that to see possible drawbacks and the convenience to remove the `$native = true` attribute.

**Comment by kikocorreoso** [comment]
Today I was testing the following operations:
```python
import random
import time

t0 = time.time()

a = list(range(0, 10000))
b = list(range(10000, 20000))
c = []
d = []
e = []

t1 = time.time()

for i in range(len(a)):
    c.append(a[i] + b[i])
    d.append(a[i] - b[i])
    e.append(a[i] * b[i])

t2 = time.time()

print(t1-t0)
print(t2-t1)
print(t2-t0)
```

<table>
 <tr>
  <td>With <code>$native=true</code> </td>
  <td>the code between t1 and t2 was around 1100ms.</td>
 </tr>
<tr>
  <td>With <code>//$native=true</code>, i.e. without the attribute <code>$native</code></td>
  <td>the code between t1 and t2 was around 1600ms.</td>
 </tr>
</table>

:-(

Any workaround to use properties in basic types? Some can be added as attributes but others must be properties.

**Comment by JamesHutchison** [comment]
Why would you want to add properties to the native types? In Python you would subclass the type and add a property to that.

**Comment by kikocorreoso** [comment]
@JamesHutchison 
From [PEP3141](https://www.python.org/dev/peps/pep-3141/) there is a *numbers* module that define all the operations that python number types (int, complex, float) should support.

I was just trying to use properties to add some of them not implemented (some packages that work with numbers use them). I thought it would be the right way in Brython.

**Comment by PierreQuentel** [comment]
Kiko,

I propose the following for types with attribute `$native` set : 
- create descriptors like ordinary methods, eg 

```javascript
$IntDict.numerator = function(self){return self}
```

- define an attribute `descriptor` for the class of the built-in type
- set it to a Javascript object indexed by the names of the descriptors

```javascript
var $IntDict = {__class__:$B.$type,
    __name__:'int',
    __dir__:$ObjectDict.__dir__,
    toString:function(){return '$IntDict'},
    $native:true,
    descriptors:{numerator:true}
}
```

I tested it with the above example, it works and should not cause performance issues, or very limited. It is implemented in the commit referenced above. Is it ok for you ?

**Comment by kikocorreoso** [comment]
@PierreQuentel 

For me it is really cool!. Thanks for taking the time.

**Comment by kikocorreoso** [comment]
This is not working!! :-(

When I define `descriptors` property in the object the instance never calls builtin `getattr` where you modified the code to take into account descriptors (#115).

I created PR #340 in order to try to fix this and could define descriptors on native types (necessary for modules like decimals, fractions, statistics,....).

**Comment by kikocorreoso** [comment]
Itertools module tests show errors with `slice` (and `range`) due to how descriptors are defined in the new implementation of this two functions in ***py_range_slice.js*** file. I modified the file in order to make the functions work.

**Comment by kikocorreoso** [comment]
Ok, I see now where it is the problem... (I think)

The original implementation to make descriptors work (b592999) was modified so the way described here is not working with the actual implementation.

With: the actual implementation:

* Descriptors defined in `complex` type works when the [key is defined between quotes](https://github.com/brython-dev/brython/blob/master/www/src/py_complex.js#L183).
* Descriptors defined in `range` and `slice` types are not working because the [keys are not enclosed by quotes](https://github.com/brython-dev/brython/blob/master/www/src/py_range_slice.js#L174).

With the included code in PR #340 seems all should be working and descriptors should be defined as stated originally in this issue.

Please, @PierreQuentel , choose what do you prefer:

* accept the #340 and define descriptors as defined originally (a little bit more verbose)
* discard #340 and fix ***py_range_slice.js*** to enclose descriptors between quotes and define descriptors as you are doing in ***py_complex.js***.
